### PR TITLE
Exercise2

### PR DESCRIPTION
**Pull request by Herzogin**
Please review, thank you!!

https://github.com/Systems-Development-and-Frameworks/CountryRoads/pull/3#pullrequestreview-526099663

### CODE REVIEW

File: webapp/.storybook/main.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+module.exports = {
+  "stories": [
+    "../src/**/*.stories.mdx",
+    "../src/**/*.stories.@(js|jsx|ts|tsx)"
+  ],
+  "addons": [
+    "@storybook/addon-links",
+    "@storybook/addon-essentials",
+    "@storybook/addon-controls",
```
Comment (ilonae) [inline webapp/.storybook/main.js:9]: I got to know that controls are inherited from `@storybook/addon-essentials`. I think you might as well use those.

File: webapp/src/components/SortListButton/SortListButton.spec.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+            wrapper.trigger("click")
+     
+           expect(wrapper.emitted().sortorder[0]).toContain("ascending")
+        })
+        test("Button emits descending",async()=> {
+            const wrapper = mount(SortListButton,{})
+            wrapper.trigger("click")
+            wrapper.trigger("click")
+           expect(wrapper.emitted().sortorder[1]).toContain("descending")
+        })
```
Comment (ilonae) [inline webapp/src/components/SortListButton/SortListButton.spec.js:33]: I think you could use a `beforeEach` here to mount the component, to then first test the click for ascending order.
In the second test you could consecutively test for an descending order.

File: webapp/src/App.vue [hunk 3]
```diff
@@ -54,35 +60,52 @@ export default {
...
-      this.news = [...this.news].sort((a,b) => (a.votes < b.votes) ? 1 : ((b.votes < a.votes) ? -1 : 0));
+    sortItems(sort="ascending") {
+      this.sortOrder = sort;
+      // if (sort=="ascending"){
+      // this.news = [...this.news].sort((a,b) => (a.votes < b.votes) ? 1 : ((b.votes < a.votes) ? -1 : 0));
+      
+      // }
+      // else{
+      //   this.news = [...this.news].sort((a,b) => (a.votes > b.votes) ? 1 : ((b.votes > a.votes) ? -1 : 0));
+      // }
```
Comment (ilonae) [inline webapp/src/App.vue:94]: Do you still need that?
If this is unnecessary, it's better to take care of code before it starts to rot.

File: webapp/src/App.vue [hunk 4]
```diff
@@ -54,35 +60,52 @@ export default {
...
+      return [...this.news].sort((a,b) => (a.votes < b.votes) ? 1 : ((b.votes < a.votes) ? -1 : 0));
+      
+      }
+      return [...this.news].sort((a,b) => (a.votes > b.votes) ? 1 : ((b.votes > a.votes) ? -1 : 0));
+      
+    }
+  },
+  mounted() {
+    // this.sortItems()
   },
```
Comment (ilonae) [inline webapp/src/App.vue:109]: Same as above.

File: webapp/src/components/List/List.spec.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+                        
+                            id: 0,
+                            title: "VueJSTest",
+                            votes: 3,
+                            },
+                    ]
+                }
+            
+            })
+            expect(wrapper.html()).not.toContain("list is empty :(")
```
Comment (ilonae) [inline webapp/src/components/List/List.spec.js:33]: I'd rather check whether `listItems` is correctly loaded, e.g. check whether `title` for example is contained.

**Review by ilonae — APPROVED** [review]

File: exercises/2/README.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+### Optional objectives
+
+In our lecture of October 21st 2020 we covered online open-source communities and git internals. These topics are not (yet) part of the current homework but will be a task in the upcoming homeworks and/or the final exam. So it's a great opportunity for you to practice the following:
+
+:rocket: Record an online pair-programming.
+- Choose someone who is not a member of your team, preferably some member of an open-source online community but another team member is fine, too.
+- Record the pair-programming to [PeerTube](https://joinpeertube.org/en/) or [YouTube](https://www.youtube.com/) and send the link to the video to us
+- Kudos if you publish the video :+1: otherwise mark the video as "unlisted"
+
+:rocket: Ensure meaningful commit messages and `git commit --amend` your commits. (This is why we learned how git compares commits and which action will change the commit id)
```
Comment (roschaefer) [inline exercises/2/README.md:90]: I like that you merge `homework` repository frequently into your own repository! :+1: 

To keep the diff in "Files changed" small for your fellow reviewer, you might want to consider to merge/rebase the branch `homework/main` into your own `main` first. If the target branch has these commits, they will be excluded from the diff here.

File: webapp/.storybook/main.js (hunk 1 above)
Comment (roschaefer) [reply webapp/.storybook/main.js:9]: `addon-essentials` is a meta package including `addon-controls`: https://storybook.js.org/docs/react/essentials/introduction

File: webapp/README.md [hunk 7]
```diff
@@ -1,4 +1,5 @@
 # VueJS
+[![Actions Status](https://github.com/Systems-Development-and-Frameworks/dirty_thirties/workflows/CI/badge.svg)](https://github.com/Systems-Development-and-Frameworks/dirty_thirties/actions)
```
Comment (roschaefer) [inline webapp/README.md:2]: Great! :+1:

File: webapp/src/App.vue [hunk 8]
```diff
@@ -9,27 +9,33 @@
...
   <div id="app">
     <h1>News List</h1>
-    <List id="li" :listItems='news'
-      v-on:delete-item='deleteItem($event)'
-      v-on:update-item='updateItem($event)'
+    <SortListButton
+    @sortorder="sortItems"
+    />
+    <List id="li" :list-items="sortedItems"
+      @delete-item='deleteItem($event)'
```
Comment (roschaefer) [inline webapp/src/App.vue:16]:
```suggestion
      @delete-item='deleteItem'
```
Same same

File: webapp/src/App.vue (hunk 3 above)
Comment (roschaefer) [inline webapp/src/App.vue:94]: @ilonae you can suggest changes with "insert a suggestion" button above :point_up: 

```suggestion
```
This way the maintainers of this repo can commit your suggestion through Github's web UI.

File: webapp/src/App.vue [hunk 9]
```diff
@@ -54,35 +60,52 @@ export default {
...
+      //   this.news = [...this.news].sort((a,b) => (a.votes > b.votes) ? 1 : ((b.votes > a.votes) ? -1 : 0));
+      // }
     }
   },
-  mounted: function() {
-    this.sortItems()
+  computed:{
+    sortedItems(){
+       if (this.sortOrder=="ascending"){
+      return [...this.news].sort((a,b) => (a.votes < b.votes) ? 1 : ((b.votes < a.votes) ? -1 : 0));
```
Comment (roschaefer) [inline webapp/src/App.vue:100]:
```suggestion
      return [...this.news].sort((a,b) => (a.votes - b.votes));
```

File: webapp/src/components/List/List.spec.js (hunk 5 above)
Comment (roschaefer) [reply webapp/src/components/List/List.spec.js:33]: Yes, testing for nothingness is really error-prone and can lead to silently passing, false-negative tests.

File: webapp/src/components/List/List.spec.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+            
+            })
+            expect(wrapper.html()).toContain("list is empty :(")
+        })
+        test("Test if list is not empty",()=> {
+            const wrapper = mount(List,
+                {
+                    propsData:
+                {
+                    listItems:[{
```
Comment (roschaefer) [inline webapp/src/components/List/List.spec.js:23]: Indentation is hard to read here. My suggestion:
```js
let listItems
describe('given some `listItems`', () => {
  beforeEach(() => {
    listItems = [{ id: 0, title: "VueJSTest", votes: 3, }]
  })

  it('renders title of each item', () => {
    const wrapper = mount(List, { propsData: { listItems }})
    expect(wrapper.text()).toContain("VueJSTest")
  })
})
```

File: webapp/src/components/List/List.vue [hunk 11]
```diff
@@ -12,21 +12,28 @@
...
+  import  ListItem from '@/components/ListItem/ListItem'
   export default {
     name: 'List',
     components: {
         ListItem
     },
-    props: ['listItems']
+    props: {
+      listItems: {
+        required: false,
```
Comment (roschaefer) [inline webapp/src/components/List/List.vue:33]:
```suggestion
```
I think that's implicit when you define `default` function.

File: webapp/src/components/ListItem/ListItem.vue [hunk 12]
```diff
@@ -9,8 +9,8 @@
 <template>
   <div id="list-item">
     <h2>{{item.title}} - ({{item.votes}})</h2>
-    <button @click="upvoteItem();$emit('update-item', item)">Upvote</button>
-    <button @click="downvoteItem();$emit('update-item', item)">Downvote</button>
+    <button @click="upvoteItem();">Upvote</button>
```
Comment (roschaefer) [inline webapp/src/components/ListItem/ListItem.vue:12]:
```suggestion
    <button @click="upvoteItem">Upvote</button>
```
If you give VueJS a function, VueJS will call the function for you. Here, you're calling the function when VueJS renders the template. Technically no big difference, but less code.

File: webapp/src/components/ListItem/ListItem.vue [hunk 13]
```diff
@@ -22,12 +22,15 @@ export default {
   name: 'ListItem',
   props: ['item'],
   methods: {
-    upvoteItem: function() {
+    upvoteItem() {
       this.item.votes++;
```
Comment (roschaefer) [inline webapp/src/components/ListItem/ListItem.vue:26]: Do not mutate props directly within the component. Mutate in the parent component and let the child component re-render.

Comment (roschaefer) [reply webapp/src/components/ListItem/ListItem.vue:26]: See here: https://vuejs.org/v2/guide/components-props.html#One-Way-Data-Flow
and here: https://vuejs.org/v2/guide/migration.html#Prop-Mutation-deprecated

Comment (roschaefer) [reply webapp/src/components/ListItem/ListItem.vue:26]: You can solve it by creating new object on-the-fly:
```js
this.$emit('update-item', { ...this.item, votes: (this.item.votes + 1) })
```
Or initialize separate reactive state:
```js
data() {
  // you have access to your properties in data()
  this.internalItem = { ...this.item }
}
```

File: webapp/src/components/SortListButton/SortListButton.spec.js [hunk 14]
```diff
@@ -0,0 +1,36 @@
+import {mount} from '@vue/test-utils'
+import SortListButton from '@/components/SortListButton/SortListButton.vue'
+
+describe (
+    "SortListButton.vue", ()=> {
```
Comment (roschaefer) [inline webapp/src/components/SortListButton/SortListButton.spec.js:5]:
```suggestion
describe ('SortListButton.vue', () => {
```

File: webapp/src/components/SortListButton/SortListButton.spec.js [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+        test("Button emits ascending",()=> {
+            const wrapper = mount(SortListButton,
+                {
+                  
+            
+            })
+            wrapper.trigger("click")
+     
+           expect(wrapper.emitted().sortorder[0]).toContain("ascending")
+        })
```
Comment (roschaefer) [inline webapp/src/components/SortListButton/SortListButton.spec.js:27]: It's not obvious to me how these two test cases are different from each other.

File: webapp/src/components/SortListButton/SortListButton.spec.js [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+        test("Button emits descending",async()=> {
+            const wrapper = mount(SortListButton,{})
+            wrapper.trigger("click")
+            wrapper.trigger("click")
+           expect(wrapper.emitted().sortorder[1]).toContain("descending")
```
Comment (roschaefer) [inline webapp/src/components/SortListButton/SortListButton.spec.js:32]: I would also be interested in the full `emitted()` here:


```suggestion
           expect(wrapper.emitted()).toEqual({ sortOrder: ['what', 'exactly']})
```

File: webapp/src/components/SortListButton/SortListButton.vue [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+        return {
+            buttontext: "ascending"
+        }
+    },
+    methods:{
+        sortlist(){
+            this.$emit("sortorder", this.buttontext)
+            //ternärer Ausdruck
+            this.buttontext = this.buttontext=="ascending"?"descending":"ascending"
+        }
```
Comment (roschaefer) [inline webapp/src/components/SortListButton/SortListButton.vue:17]: This is obviously a computed property:

```suggestion
        sortlist(){
            this.$emit("sortorder", this.buttontext)
        }
      },
      computed: {
        buttontext() {
            return this.buttontext === 'ascending' ? 'descending' : 'ascending'
        }
```

File: webapp/src/stories/Button.stories.js [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+Large.args = {
+  size: 'large',
+  label: 'Button',
+};
+
+export const Small = Template.bind({});
+Small.args = {
+  size: 'small',
+  label: 'Button',
+};
```
Comment (roschaefer) [inline webapp/src/stories/Button.stories.js:39]:
```suggestion
```
Remove unrelated code

**Review by roschaefer — APPROVED** [review]
Nice

Nice, well done! Here's what I've found:

:star: Setup linter
:star: Setup jest
:star: Feature: Toggle ascending/descending order
:star: Feature: Empty state
:star: Co-locate stories, tests and component code
:star: Setup build server
:star: Setup storybook
:star: Storybook interactively controllable
:star: Storybook passes a11y tests
:star: Give a code review

That makes 10/10 :star: congratulations!

Here some more general suggestions:
1. Please remove unrelated code
2. You could update target branch `main` first before you create your PR
3. You could improve your test case descriptions:
```
 PASS  src/components/List/List.spec.js
  List.vue
    ✓ Test if List is empty (27ms)
    ✓ Test if list is not empty (10ms)
```

I would prefer something like this:
```
 PASS  src/components/List/List.spec.js
  List.vue
    given no listItems
      ✓ displays empty state (27ms)
    given some listItems
      ✓ displays each `listItem.title` (10ms)
```
